### PR TITLE
AP_Proximity: Return maximum distance and minimum distance with Proximity type

### PR DESCRIPTION
**Pull request by muramura**
I will make it possible to set the maximum and minimum values of A2 series by config parameter.
Also consider the correspondence to A1M8 and A3M1 of the same protocol.
I have A2 series, A2M4, A2M6(-R3, -R4), A2M8 kind in the A2 series.
For each type, the minimum and maximum values are different.
This driver sets the maximum and minimum values of A2M6-R3 to fixed values.

and-
I initialize the class variable in the constructor.
This variable is initialized to activation level.
Therefore, I do not think you need to initialize it with the constructor.

### CODE REVIEW

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 1]
```diff
@@ -148,6 +148,24 @@ const AP_Param::GroupInfo AP_Proximity::var_info[] = {
...
+    AP_GROUPINFO("_DISTMAX_M", 19, AP_Proximity, _distance_max[0], 16.00f),
+
+    // @Param: __DISTMIN_M
+    // @DisplayName: minimum distance (in meters) of sensor
+    // @Description: minimum distance (in meters) of sensor
+    // @Units: m
+    // @Range: 0.00 25.00
+    // @Increment: 0.01
+    // @User: Standard
+    AP_GROUPINFO("__DISTMIN_M", 20, AP_Proximity, _distance_min[0], 0.20f),
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp]: double underscore I think and normally we don't include the units in the name..

**Comment by rmackay9** [comment]
Is this going to work ok with the MAV backend?  I know it gets it's maximum distance from the mavlink message..

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 2]
```diff
@@ -171,6 +189,24 @@ const AP_Param::GroupInfo AP_Proximity::var_info[] = {
     // @Range: -180 180
     // @User: Standard
     AP_GROUPINFO("2_YAW_CORR", 18, AP_Proximity, _yaw_correction[1], PROXIMITY_YAW_CORRECTION_DEFAULT),
+
+    // @Param: _DISTMAX_M
+    // @DisplayName: maximum distance (in meters) of sensor
```
Comment (amilcarlucas) [inline libraries/AP_Proximity/AP_Proximity.cpp]: remove the "(in meters)" from @DisplayName . You can leave it in @Description

File: libraries/AP_Proximity/AP_Proximity.h [hunk 3]
```diff
@@ -78,6 +78,10 @@ class AP_Proximity
     uint8_t get_orientation(uint8_t instance) const;
     int16_t get_yaw_correction(uint8_t instance) const;
 
+    // return sensor maximum and minimum distance (in meters)
+    float get_distance_max(uint8_t instance) const;
```
Comment (amilcarlucas) [inline libraries/AP_Proximity/AP_Proximity.h]: rename to get_distance_max_m() and get_distance_min_m()

File: libraries/AP_Proximity/AP_Proximity.h [hunk 4]
```diff
@@ -157,6 +161,8 @@ class AP_Proximity
     AP_Int16 _yaw_correction[PROXIMITY_MAX_INSTANCES];
     AP_Int16 _ignore_angle_deg[PROXIMITY_MAX_IGNORE];   // angle (in degrees) of area that should be ignored by sensor (i.e. leg shows up)
     AP_Int8 _ignore_width_deg[PROXIMITY_MAX_IGNORE];    // width of beam (in degrees) that should be ignored
+    AP_Float _distance_max[PROXIMITY_MAX_INSTANCES];  // maximum distance (in meters) of sensor
```
Comment (amilcarlucas) [inline libraries/AP_Proximity/AP_Proximity.h]: Add _m at the end of the variable name

File: libraries/AP_Proximity/AP_Proximity.cpp (hunk 1 above)
Comment (muramura) [reply libraries/AP_Proximity/AP_Proximity.cpp]: @rmackay9 san.. Comments thanks. I'll fix it.

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 5]
```diff
@@ -118,13 +115,13 @@ void AP_Proximity_RPLidarA2::update(void)
...
 float AP_Proximity_RPLidarA2::distance_max() const
 {
-    return 16.0f;  //16m max range RPLIDAR2, if you want to support the 8m version this is the only line to change
+    return frontend.get_distance_max(state.instance);  //16m max range RPLIDAR2, if you want to support the 8m version this is the only line to change
 }
 
 // get minimum distance (in meters) of sensor
 float AP_Proximity_RPLidarA2::distance_min() const
 {
-    return 0.20f;  //20cm min range RPLIDAR2
```
[the remark this answers is not in the document]
Comment (muramura) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp]: @amilcarlucas san, @rmackay9 san.. 
I add 4 kinds of A2M4, A2M8, A1M8, A3M1 to the configuration parameter "Proximity type".
I will try to return the maximum and minimum values with this type of value.

**Comment by muramura** [comment]
@rmackay9 san, @amilcarlucas san...
I changed the maximum and minimum values by looking at Proximity type.

**Comment by amilcarlucas** [comment]
This can probably be closed, now that #23749 has been merged.
Or are there features of this PR that you are missing on #23749 ?

**Pull request closed without merging**